Update shared work UI

sfentress opened the pull request:
This updates some styles, and creates a split-panel left navigation that shows student names at the top and a space for comments at the bottom (which will be in the next PR).

The split-panel was made without dependencies, as it is pretty straightforward. (There were some React libraries but I wasn't able to style them the way that was requested.)

The new style (with working comments) can be seen at http://lara-sharing-plugin.concord.org/branch/post-comments/demo.html

dougmartin reviewed (approved):
Looks good :+1:

kswenson commented on src/components/split-pane.tsx:
```diff
@@ -0,0 +1,88 @@
+    separatorYPosition.current = null;
+  };
+
+  React.useEffect(() => {
```
In addition to performance, Doug's suggestion would also allow you to use capturing handlers, which would reduce the chance that anyone else could grab the events.

kswenson commented on src/components/split-pane.tsx:
```diff
@@ -0,0 +1,85 @@
+    document.addEventListener("mousemove", onMouseMove, {capture: true});
+    document.addEventListener("touchmove", onMouseMove, {capture: true});
+    document.addEventListener("mouseup", onMouseUp, {once: true});
+    document.addEventListener("touchend", onMouseUp, {once: true});
```
Interesting! I don't think I've ever used `once: true` before, but it makes perfect sense. Is there a reason you didn't make the `mouseup`/`touchend` handlers `capture: true` as well?

sfentress replied:
It seemed less relevant if someone else also captured mouseUp. I feel like in general as few events as possible should be captured if not needed, as there is a purpose to bubbling. But I used them in the `mouseMove` case in particular because I was trying to get rid of an annoying bug on mobile where the drag events were also captured by the page beneath the pop-up window, which drags up and down. I haven't actually been able to lick the bug, but it seemed that (totally subjectively) after I was capturing, plus also using `stopImmediatePropagation`, that touch-handling worked slightly better.

kswenson replied:
If someone else gets the `mouseup` first, that someone might also `stopPropagation()`, in which case this listener would never see the event and the drag would continue indefinitely, no?

sfentress replied:
That scenario is precisely the reason that I think handlers should be extremely limited about using things like `stopPropagation` unnecessarily. I'm also unsure why anyone would stop a mouseUp from propagating.

Also, unless my understanding is wrong, at this element we are in the innermost nest of the elements that are capturing and bubbling the event (nothing further is going to nest in this handle element that someone else will control). The capture phase goes from the outermost to the innermost, and then it bubbles back out again. So, unless I'm mistaken, there's nothing that can come between the capture event hitting this element and the bubble event hitting this element, as this is where the "bounce" happens. If an outer element stopped propagation during capture phase, we'd never see the event, and if it stopped propagation during the bubble phase it wouldn't matter.

All that said, if there is some reason we're trying to switch to using `capture` on all mouse events, I'm all ears, knowing little about this, but until now I haven't seen that much as a pattern in any documentation of handling mouse events.

kswenson replied:
I agree it's a subtle area and at this point we're splitting hairs, so I'm fine with whatever you think makes sense. That said, my thinking has been that since these handlers are being installed on `document`, pretty much every other DOM element will see these events before they bubble up to here and so `capture: true` is the difference between being at the beginning of the line versus the end of the line. Since during the drag we believe that this code should take priority over whatever else is going on, my pattern has been to capture both move and up and then remove the handlers on up (as you're doing).

sfentress replied:
> That said, my thinking has been that since these handlers are being installed on `document`...

D'oh, totally refutes my point about the bubble order. I still had it in my head that the listener was on the target element. That's a good point then, any other element capturing *would* block it.

kswenson replied:
In practice there's another wrinkle in that React ≤ 16 attaches its event handlers at the `document` level as well which complicates the thinking about the order, but React 17+ attaches handlers at the root DOM element, which would make the capture scenario even more likely, so in the end I'm not sure thinking about React helps any more than the general rule that capturing helps guarantee that you get the event.

kswenson commented on src/components/split-pane.tsx:
```diff
@@ -0,0 +1,85 @@
+    document.removeEventListener("mousemove", onMouseMove);
+    document.removeEventListener("touchmove", onMouseMove);
+    document.removeEventListener("mouseup", onMouseUp);
+    document.removeEventListener("touchend", onMouseUp);
```
From [MDN](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener), I don't think you need to explicitly remove the listeners added with `once: true` because they should be removed automatically.

sfentress replied:
You're right, I figured it couldn't harm, but it is probably unnecessary.

sfentress commented:
@kswenson An issue I was having which you might be interested in trying to figure out the best way to have correctly handled, but no worries if not, my current solution is working:

I spent a whole bunch of time trying to fix an issue that was introduced when I moved where the `mouseMove` handlers get attached in e384d95.

`onMouseMove` was referencing the state property `topPosition`, which it itself was setting on every call. Previously, when the handler was getting attached in `useEffect`, this worked as expected, and the function always correctly referenced the latest value of `topPosition`.

Once I switched to attaching the handler only on `mouseDown` in commit e384d95, the value of `topPosition` seemed to have gotten locked in a closure and was not being updated. I initially assumed that this was only happening within the current-drag cycle, and made a change to the math to compensate, which meant that the current drag correctly moved the splitter. I hadn't noticed, however, that a second drag would always reset the splitter to the initial location. This was because `onMouseDown` was never getting the new `topPosition`, even if I set it in `onMouseDown`.

If I console.logged the value of `topPosition` in both `onMouseDown` and `onMouseMove`, the former would show the freshest value, while the latter would permanently show only the very first value.

From [this SO answer](https://stackoverflow.com/a/55906738/1231271) it seemed that the could be solved with `useCallback` and passing `topHeight` as a dependency, but this did not work.

I ended up solving it by duplicating the current value of `topHeight` into a `ref` in `mouseDown`, and only reading from that ref in `mouseMove`. The ref correctly referenced the most recent value.

Any thoughts as to whether there was a more idiomatic hooks way to have solved this?

sfentress commented:
Ah, I found my solution was identical to [this SO answer](https://stackoverflow.com/a/55156813/1231271).

kswenson commented:
Stale closures is a recurring gotcha with functional components/hooks. Putting the value in an auto-updating ref is (as you discovered) a typical solution. The form of the solution in the SO answer you referenced could be encapsulated as:
```typescript
const useCurrent = <T>(value: T) => {
  const refValue = useRef(value);
  useEffect(() => {
    refValue.current = value;
  });
  return refValue.current;
};
```

I have successfully used a simpler version:
```typescript
const useCurrent = <T>(value: T) => {
  const refValue = useRef(value);
  refValue.current = value;
  return refValue.current;
};
```

I've never found a good explanation for why waiting until after the render (via `useEffect()`) to update the value is preferable to just setting the value directly.
